### ISSUE TITLE
Meta issue: type system extensions

### ISSUE THREAD

**Issue by evancz** [elm-lang/elm-compiler#1039]
**This issue is to aggregate features that have been requested. If you want to add something open a *separate* issue, this is just for centralizing things.**

The ML-family has a rich history, with many different languages and many different type system features. The most commonly requested type system features are:

  * Type classes (see #38, one of the first issues for Elm!)
  * Higher-kinded polymorphism (HKP)
  * Rank N polymorphism (see #238)

These requests usually come from folks coming from Haskell who want Elm to be Haskell.

There are a variety of ways to get roughly the same sort of "genericness" in your code.

  * type classes - the most widely known approach, used in Haskell
  * HKP + RankN - read [Scrap your Type Classes](http://www.haskellforall.com/2012/05/scrap-your-type-classes.html) to get the basic idea and pretend the syntax is nicer.
  * HKP + RankN + Implicit Arguments - [On the Bright Side of Type Classes](https://lirias.kuleuven.be/bitstream/123456789/304985/1/icfp001-Devriese.pdf) describes how this works in Agda
  * module functors - this is the approach used in SML and OCaml, which both have very sophisticated module systems. It lets you do tons of stuff above and beyond these other approaches. It looks like OCaml may add some implicit resolution feature as well.

I think all of these approaches are compelling, and since the very beginning of Elm, it has not become clear which is "the right choice" for Elm. It is also true that if you go too crazy adding this stuff, you probably can never un-add it.

I think [this talk](https://www.youtube.com/watch?v=oYk8CKH7OhE) is probably one of the best explanations I have given of the design philosophy that led to the "wait and see" approach we are taking on these topics. You can also search the archives of the elm-discuss to find the numerous and often tumultuous threads about these features.

**Comment by JoeyEremondi** [comment]
My current leaning is to "HKP + RankN". Nothing is implicit, it's really no more verbose than our style of having everything qualified by module, and it avoids things like Monomorphism restriction, orphan instances, conflicting instances, breaking inference, etc.

But it's really hard to evaluate these things without seeing what code using them would look like.

Something to consider here, is that there two main people who write Elm code:

* Library maintainers, who want to make generic code
* Application developers, who want to use libraries to make their life easier.

Some of these features, like RankN types, are confusing for newcomers. But they will almost never write such a things in an enclosed application. People writing generic libraries, however, can and should understand how Elm's type system works.

Perhaps this could be codified somehow, with a flag to enable advanced features? Haskell's extension system is messy, but it does allow you to avoid accidentally coming across an advanced feature you don't understand.

> These requests usually come from folks coming from Haskell who want Elm to be Haskell.

This is partially true, but in many cases it's not. People want a ready-to-go solution for generic programming. There's a huge difference between wanting fancy types, and wanting to be able to store arbitrary types in a `Dict`, or not have 10 different versions of `forM` (or whatever we'd call it in Elm) floating around in *-extras libraries for each type that implements `andThen`.

**Comment by rtfeldman** [comment]
@JoeyEremondi I have some thoughts on this, but I believe the intention is for this issue is to open separate issues for specific ideas. Would you mind opening an issue like "HKP + RankN with no implicits" so we can discuss there and @evancz can link to it from this issue?

**Comment by rgrempel** [comment]
Here's a meta-comment on a meta-issue. 

I think it's clear that whatever Elm eventually does with respect to type system extensions will need to emerge from concerns which are "native" to Elm, so to speak -- that is, it will need to (a) solve problems that are actually faced by Elm programmers and (b) make it easier to use idioms that Elm wants to encourage.

Whenever this happens, I think that the first place to look for use cases would be the `Basics` module, and in particular `comparable`, `appendable` and `number`. My reasoning is that clearly Elm has already determined that these are "fancy types" that do solve real Elm problems, and do make it easier to use idioms that Elm wants to encourage. At the moment, I presume that these types are special-cased in one way or another by the compiler. So, it seems as though one sensible question to ask about any proposal for type system extensions would be: could this extension implement `comparable`, `appendable` and `number` as a general-case, rather than as special-cases?

I do have other ideas for Elm-native use cases for fancy types. However, at the moment I'm still working at pushing those use cases as far as possible within the current type system, so I won't mention them yet. That is, I'm not sure yet if those other use cases absolutely need fancy types.

**Comment by sindikat** [comment]
Gabriel Gonzalez just posted [an article](http://www.haskellforall.com/2015/10/explicit-is-better-than-implicit.html), where he claims:

> Many of the limitations associated with Haskell type classes can be solved very cleanly with lenses.

Is this relevant for the thread?

**Comment by Fresheyeball** [comment]
Just a suggestion on naming should Elm go the typeclass route. 

```elm
interface Monoid a where
   (<>) : a -> a -> a
     id : a

type wrap Sum = Int 

implement Monoid Sum where
   (<>) = (+)
     id = 0

implement Monoid (List a) where
  (<>) = (++)
    id = []
```

I've found it challenging to explain typeclasses to JavaScript or OOP developers, in part due to the keywords Haskell chose. The above is an attempt to be more intuitive.

**Comment by mgold** [comment]
I'm not sure that's the right route for Elm... but *damn* that would have been so much better in Haskell. Especially with a better name than `Monoid`. (See also [this essay](https://medium.com/@Max_Goldstein/every-language-is-invented-c67d733ec01d) I wrote the other day.)

**Comment by periodic** [comment]
The fact that we have comparable as an compiler special-case shows that there is need for some solution.  I feel it would be worth coming up with a more general one than special-cases.

HKP and at least Rank-2 types seem like generic things to add.  They are more about mathematical descriptions of types than a language implementation detail.  Implicit arguments or typeclasses feel more like syntactic sugar.  They really tie you down to specific implementations and they require more decisions to be made about the language. I think the safest route is to add HKP and Rank-N types first and deal with the awkward syntax until a decision is made regarding how make the implementation look good.

Does adding either of those constrain the compiler significantly?

I just started with Elm, after following the announcements for some time and getting interested.  I was really enjoying it up to the point that I wanted to use a custom type as a Dict key and started doing some monadic programming.  I'll admit to having a very Haskell-oriented way of thinking, but I'm open to the idea of using records for these things.  As it is I'm getting along by limiting myself to one instance per module so that I can define my operators without conflicts.  I'm used to making compromises as I work in C++ fairly often.

**Comment by rtfeldman** [comment]
> The fact that we have `comparable` as an compiler special-case shows that there is need for some solution. I feel it would be worth coming up with a more general one than special-cases.

@periodic Welcome to Elm! Variations on your comment are something of a rite of passage when arriving at Elm from Haskell, to the point where a GitHub autoresponder would not be inappropriate. :wink: 

For background, Evan has master-level expertise in HKP and RankN, has voraciously consumed academic literature on various ways such features are implemented in other languages, has spent well over 200 hours (that's a very conservative number; I'm sure it's more than that) discussing and debating **only this exact question** with members of the Elm community and academic researchers, and has concluded that [adding them to Elm anytime soon is not a good idea](https://github.com/elm-lang/elm-compiler/issues/38#issuecomment-135928559).

Anyway, the point is that the community is generally sick of with discussing this topic and has generally moved on to building stuff. :smile:

**Comment by mietek** [comment]
@rtfeldman: I appreciate your comment, as I was about to raise the same issue.

**Comment by mgold** [comment]
Yes, good response @rtfeldman. I'll just add that there is not one [but three special cases](https://github.com/elm-guides/elm-for-js/blob/master/How%20to%20Read%20a%20Type%20Annotation.md#constrained-types).

**Comment by JoeyEremondi** [comment]
It's also important to remember that we don't need these fancy types or hkp
or implicits to do custom keys in dictionaries. We just need someone to
implement dictionaries that take a comparison as an argument.

**Comment by mgold** [comment]
@JoeyEremondi The core Dict is written in Elm, so whoever is so inclined can copy that and publish a third-party library version that does that. You could also add an `equals` function while you're at it...

**Comment by JoeyEremondi** [comment]
Yeah, I've got a draft of this started, I just need to get around to
actually writing it.

**Comment by periodic** [comment]
I misunderstood.  I read the closed bugs and thought this bug was the one about tracking the meta issue and figuring out what to actually *do* going forward.  It sounded to me like Evan might have been actually looking at options.  I see however that it's not an appropriate place for discussion just based on the bolded statement at the top.

This thread seemed like the place where things might be moving towards implementing something to generalize the special casing.  I got a little too excited.

Back to reading the other threads and experimenting with work-arounds.

**Comment by mgold** [comment]
No worries, happy experimenting!

**Comment by tonymorris** [comment]
> has master-level expertise in HKP and RankN … and has concluded that …

Holy shit. Not again.

**Comment by cjhowe** [comment]
I think fixing this would clean up a lot of really messy code I've been writing recently. Started using Elm for work, so I can provide a lot of real world examples if it helps.

**Comment by JoeyEremondi** [comment]
It would help very much!

**Comment by Fresheyeball** [comment]
@cjhowe which extensions specifically do you think would clean up your code?

**Comment by tiziano88** [comment]
I think Type Classes are going to be needed in one way or another; perhaps we can discuss issues (if any) with the way Haskell does this, and decide whether to start thinking about a similar approach for Elm, or whether we would like to change anything? Whatever we do, it would be good to keep in mind the use case for HKTs when designing Type Classes, so that we can easily build further abstractions later on on top of this.

**Comment by JoeyEremondi** [comment]
@tiziano88 Those are definitely not going to happen without concrete examples of what code you have in real Elm projects that is bad, that would be made better by typeclasses, that outweigh their complexity.

**Comment by tiziano88** [comment]
@JoeyEremondi definitely; I am working on a protobuf compiler for Elm (http://github.com/tiziano88/elm-protobuf for the curious -- not quite ready for general use yet) and there are a number of places where a way of saying "all those types share some of the same underlying functionality" would make things much cleaner. I will follow up here with concrete examples when I have them.

**Comment by mgold** [comment]
Please do follow up with examples, preferably on the [mailing list](https://groups.google.com/forum/#!forum/elm-discuss.
